airbyte-integrations/connectors/source-blob/source_blob/fileclient.py

```diff
@@ -0,0 +1,263 @@
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+
+from abc import ABC, abstractmethod
+from datetime import datetime
+import json
+import traceback
+from boto3 import session as boto3session
+import smart_open
+from azure.storage.blob import BlobServiceClient
+from botocore import UNSIGNED
+from botocore.config import Config
+from google.cloud.storage import Client as GCSClient
+from google.oauth2 import service_account
+from airbyte_cdk.logger import AirbyteLogger
+
+
+# NOTE: this is based off the URLFile class from source-file's client.py
+
+
+class ConfigurationError(Exception):
+    """Client mis-configured"""
+
+
+class PermissionsError(Exception):
+    """User doesn't have enough permissions"""
+
+
+class FileClient(ABC):
+    """ TODO docstring
+    Manages accessing a file using smart_open. Child classes implement provider specific logic.
+    """
+
+    def __init__(self, url: str, provider: dict):
+        self._url = url
+        self._provider = provider
+        self._file = None
+        self.logger = AirbyteLogger()
+
+    def __enter__(self):
+        return self._file
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.close()
+
+    @property
+    @abstractmethod
+    def url(self) -> str:
+        """Convert URL to remove the URL prefix (scheme)
+        :return: the corresponding URL without URL prefix / scheme
+        """
+
+    @property
+    def fullurl(self):
+        return f"{self.storage_scheme}{self.url}"
+
+    @property
+    @abstractmethod
+    def storage_scheme(self) -> str:
+        """Convert Storage Names to the proper URL Prefix
+        :return: the corresponding URL prefix / scheme
+        """
+        # storage_name = self._provider["storage"].upper()
+        # parse_result = urlparse(self._url)
+        # if storage_name == "HTTPS":
+        #     return "https://"
+        # elif storage_name == "SSH" or storage_name == "SCP":
+        #     return "scp://"
+        # elif storage_name == "SFTP":
+        #     return "sftp://"
+        # elif storage_name == "WEBHDFS":
+        #     return "webhdfs://"
+        # elif storage_name == "LOCAL":
+        #     return "file://"
+        # elif parse_result.scheme:
+        #     return parse_result.scheme
+        # self.logger.error(f"Unknown Storage provider in: {self.fullurl}")
+        # return ""
+
+    @property
+    @abstractmethod
+    def last_modified(self) -> datetime:
+        """ TODO Docstring """
+
+    def close(self):
+        if self._file:
+            self._file.close()
+            self._file = None
+
+    def open(self, binary=False):
+        self.close()
+        self._file = self._open(binary=binary)
+        return self
+
+    @abstractmethod
+    def _open(self, binary):
+        """TODO docstring"""
+        # mode = "rb" if binary else "r"
```

Comment:
Copied from existing file connector, left in commented out for reference but will most likely remove